### PR TITLE
refactor: generalize universes for colimits and limits in Type

### PR DESCRIPTION
**Pull request by joelriou**
In the category `Type u` and other concrete categories, there are limits and colimits indexed by categories `J : Type v` when `[UnivLE.{v, u}]`. As the universe for morphisms in `J` is irrelevant, there is no longer any assumption that `J` is a small category.

---

The definition of the standard colimit cocone in `Type u` now uses `Shrink` (see `CategoryTheory.Limits.Types`), which broke some constructions of colimits which relied a little too much on the definitional properties of this cocone. While fixing these constructions, some proofs have been cleaned up. Assumptions on universes have been generalized when appropriate.

The same should be done for the standard limit cone, but I would rather attempt at this in a separate PR...

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jcommelin** [comment]
How does this PR mesh with the planned refactor to use `univLE`? This file is still using `TypeMax`, although the diff in the PR doesn't seem to touch it really. I haven't thought through the question myself.

**Comment by joelriou** [comment]
My PR is orthogonal to the `UnivLE` refactor, which seems to be done already for limits of types in this file, but not for colimits, where the colimit cocone is constructed only in `TypeMax`.

Ideally, under `[UnivLE.{v, u)]` we should have limits and colimits in `Type u` indexed by categories `J` such that `J : Type v` and `[Category.{w} J]`. Currently, there is the irrelevant assumption that `w = v`. My PR only removes any assumption on `w`, but maybe I should try to also adapt the construction of colimits under the assumption `UnivLE.{v, u}` rather than keeping only the `TypeMax` version.

File: Mathlib/Algebra/Category/Ring/FilteredColimits.lean [hunk 1]
```diff
@@ -179,26 +179,26 @@ section
...
-      (F ⋙ forget₂ CommSemiRingCat CommMonCat.{max v u}) with }
+      (F ⋙ forget₂ CommSemiRingCat CommMonCat) with }
 set_option linter.uppercaseLean3 false in
 #align CommSemiRing.filtered_colimits.colimit_comm_semiring CommSemiRingCat.FilteredColimits.colimitCommSemiring
 
 /-- The bundled commutative semiring giving the filtered colimit of a diagram. -/
-def colimit : CommSemiRingCat.{max v u} :=
-  CommSemiRingCat.of <| R.{v, u} F
+def colimit : CommSemiRingCat.{u} :=
+  CommSemiRingCat.of <| R  F
```
Comment (alreadydone) [inline Mathlib/Algebra/Category/Ring/FilteredColimits.lean:201]:
```suggestion
  CommSemiRingCat.of <| R F
```

File: Mathlib/CategoryTheory/Limits/Types.lean [hunk 2]
```diff
@@ -314,155 +308,173 @@ def Quot.Rel (F : J ⥤ TypeMax.{v, u}) : (Σ j, F.obj j) → (Σ j, F.obj j) 
...
-theorem Colimit.w_apply' {F : J ⥤ Type v} {j j' : J} {x : F.obj j} (f : j ⟶ j') :
-    colimit.ι F j' (F.map f x) = colimit.ι F j x :=
-  congr_fun (colimit.w F f) x
-#align category_theory.limits.types.colimit.w_apply' CategoryTheory.Limits.Types.Colimit.w_apply'
-
-@[simp]
-theorem Colimit.ι_desc_apply' (F : J ⥤ Type v) (s : Cocone F) (j : J) (x : F.obj j) :
-    colimit.desc F s (colimit.ι F j x) = s.ι.app j x :=
-  congr_fun (colimit.ι_desc s j) x
-#align category_theory.limits.types.colimit.ι_desc_apply' CategoryTheory.Limits.Types.Colimit.ι_desc_apply'
```
Comment (alreadydone) [inline Mathlib/CategoryTheory/Limits/Types.lean:415]: Should we change these to `#noalign`s, or we don't care?

File: Mathlib/Geometry/RingedSpace/PresheafedSpace/HasColimits.lean [hunk 3]
```diff
@@ -109,7 +109,7 @@ def componentwiseDiagram (F : J ⥤ PresheafedSpace.{_, _, v} C) [HasColimit F]
 set_option linter.uppercaseLean3 false in
 #align algebraic_geometry.PresheafedSpace.componentwise_diagram AlgebraicGeometry.PresheafedSpace.componentwiseDiagram
 
-variable [HasColimitsOfShape J TopCat.{v}]
+variable [UnivLE.{u', v}]
```
Comment (alreadydone) [inline Mathlib/Geometry/RingedSpace/PresheafedSpace/HasColimits.lean:112]: Why this change? I think it's now less general?

**Comment by alreadydone** [comment]
There probably isn't much benefit to be garnered from this approach, so closing.

**Review by alreadydone — COMMENTED** [review]
I like the generalizations made in this PR, but there are now merge conflicts. Also, I think many `UnivLE.{v,u}` can be replaced by `Small.{u} J`, and `UnivLE` is only really needed in `HasLimits`/`HasLimitsOfSize` instances, and `Small` is enough for `HasLimitsOfShape`. Do you think this is a good idea, and would you mind doing it?

**Comment by joelriou** [comment]
I am closing this as @alreadydone may have found a better approach.
